Syntactic region for ppx nodes

andreypopp opened the pull request:
This is an attempt to fix the highlighting for `[@@bs.module ...]` (defined by BuckleScript) in external declarations.

This defines a syntactic region `ocamlPpx` between `[@@`/`]` and `[@`/`]` which allows only `ocamlString` inside.

I linked the `ocamlPpx` to `Comment` but this is debatable.

This is how it looked before:
<img width="746" alt="screen shot 2018-04-23 at 00 16 40" src="https://user-images.githubusercontent.com/30594/39099970-9bb02f6e-468b-11e8-9294-6013f233c9aa.png">
... and now this is how it looks:
<img width="746" alt="screen shot 2018-04-23 at 00 15 44" src="https://user-images.githubusercontent.com/30594/39099973-a5ce3eb4-468b-11e8-8602-60750bbaba6b.png">

I'm not into Vim syntax highlighting so I apologise if this fix is silly, at least this highlights (pun intended) the issue.

hcarty commented:
Is it possible that the issue is caused by something else?  A piece of made-up code seems to highlight properly for me when using the current vim-ocaml
<img width="445" alt="image" src="https://user-images.githubusercontent.com/199147/39216816-6e7d9a36-47da-11e8-9dc2-46a0b9cb76f2.png">

hcarty commented:
Ah, apologies @andreypopp - your example shows the same issue here.  It looks like the highlighting is getting confused about the keywords embedded in the `bs.*` annotations.

andreypopp commented:
Thanks, @copy and @hcarty. Working on improvements.

hcarty commented on syntax/ocaml.vim:
```diff
@@ -109,6 +109,10 @@ endif
 
+"" PPX nodes
+
+syn region ocamlPpx matchgroup=ocamlPpxEncl start="\[@@\w\+\(.\w\+\)*" contains=TOP end="\]"
```
`start` should probably use something like `@\{1,3\}` to consolidate this down to one line and support `[@@@attr]`.

hcarty replied:
More concretely:
```
start="\[@\{1,3\}\w\+\(.\w\+\)*"
```

hcarty replied:
Incorporating @copy's `.` fix:
```
start="\[@\{1,3\}\w\+\(\.\w\+\)*"
```
which avoids highlighting syntax like `[@attr*foo]`.

copy commented on syntax/ocaml.vim:
```diff
@@ -109,6 +109,10 @@ endif
 
+"" PPX nodes
+
+syn region ocamlPpx matchgroup=ocamlPpxEncl start="\[@@\w\+\(.\w\+\)*" contains=TOP end="\]"
```
The `.` needs to be escaped.

copy commented:
This looks good, but we should split the the `[@@` part and the following identifier into two groups, so they can by coloured differently (in your theme it looks alright, but in my theme it looks a bit odd that syntax elements and names have the same colour).

Something like this should work:

```
syn region ocamlPpx matchgroup=ocamlPpxEncl start="\[@" contains=TOP end="\]"
syn match ocamlPpxIdentifier /\(\[@\)\@<=\w\+\(\.\w\+\)*/
```

andreypopp commented:
Thanks for the detailed feedback!

I fixed the mistakes in the pattern and added the separate syntax for ppx attr-id as requested.

This is how it looks:
<img width="752" alt="screen shot 2018-04-27 at 21 54 30" src="https://user-images.githubusercontent.com/30594/39380058-ec233dbe-4a65-11e8-81d0-d86eb552964a.png">

This is the screenshot with the presence of `highlight link ocamlPpxIdentifier SpellBad` which tests the customisation of highlighting the ppx's attr-id:

<img width="752" alt="screen shot 2018-04-27 at 21 55 20" src="https://user-images.githubusercontent.com/30594/39380094-0f5ea16a-4a66-11e8-8704-075504b28945.png">

This is the screenshot with the presence of `highlight link ocamlPpxEncl SpellBad` which tests the customisation of highlighting the ppx's enclosing tokens:

<img width="752" alt="screen shot 2018-04-27 at 21 58 53" src="https://user-images.githubusercontent.com/30594/39380133-31dff518-4a66-11e8-99f2-347194978547.png">

copy commented:
@andreypopp Thanks, cherry-picked onto master. I did change the commit and remove the `HiLink ocamlPpxIdentifier Keyword` as it is a change from the current style, looks slightly worse (in my opinion) and [at least one person agrees with me](https://irclog.whitequark.org/ocaml/2018-04-28#1524876400-1524876638;).

andreypopp commented:
Thank you!
